goose/providers: add Venice.ai private open-source LLM

faces-of-eth opened the pull request:
## Description

Implements Venice.ai as a new model provider with support for various models including:

- llama-3.3-70b (65K context)
- qwen-2.5-coder-32b for code optimization
- qwen-2.5-vl for vision capabilities
- Additional specialized models

The implementation includes:

- Live model list fetching for new model releases
- Full provider API integration
- UI configuration settings
- Improved error handling for model fetching
- Support for Venice.ai's authentication system

## Changes

- Added Venice.ai provider implementation
- Updated provider factory to include Venice.ai
- Added UI components for Venice.ai configuration
- Improved model fetching and error handling
- Added Venice.ai models to the hardcoded list

## Testing

Tested with multiple Venice.ai models including general purpose models and specialized ones for code and vision tasks.

Closes #2239

<img width="701" alt="Screenshot 2025-04-17 at 3 41 13 PM" src="https://github.com/user-attachments/assets/0c3c9036-f32b-4553-87e9-d99bb3420600" />
<img width="706" alt="Screenshot 2025-04-17 at 3 40 09 PM" src="https://github.com/user-attachments/assets/9d52e6e6-7e5e-4687-b0c1-ce2eccf7092a" />



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210278349818735

faces-of-eth committed:
feat(providers): add Venice.ai provider integration (Closes #2239)

lily-de commented:
Hey -- you don't need to update anything in the ui. Can remove those changes and if you've properly setup the provider metadata in the rust code for Venice AI it will show up as an option

faces-of-eth committed:
feat(venice): update provider implementation and UI components

faces-of-eth committed:
Merge branch 'main' into issue-2239-venice-ai

lily-de commented on ui/desktop/src/components/settings/models/hardcoded_stuff.tsx:
this is no longer needed fyi -- if you correctly added provider metadata for Venice you should see it in the UI. Lmk if you don't

lily-de commented:
LGTM -- no need to remove those changes from hardcoded stuff

angiejones commented:
dang almost there! there's a [build error](https://github.com/block/goose/actions/runs/14936613858/job/41980449023?pr=2252#step:10:585)

faces-of-eth committed:
Fixed test failures

michaelneale commented:
if you run `cargo clippy -- -D warnings ` or tell goose to run it and fix it should get it over the line

faces-of-eth committed:
fix(config): ensure atomic file operations in set_param

The test_concurrent_extension_writes test was failing due to a race condition
in the config file operations. Multiple threads could read the file
simultaneously, each getting the same initial state, and then overwrite each
other's changes when writing back.

The fix ensures atomicity by:
1. Acquiring an exclusive file lock before reading
2. Holding the lock throughout the entire read-modify-write cycle
3. Using a single file handle for the entire operation

This prevents data loss in concurrent scenarios while maintaining the existing
API contract. The fix is isolated to set_param and doesn't affect the behavior
of other public methods.

Test: test_concurrent_extension_writes now passes consistently

alexhancock commented:
@faces-of-eth There is some discussion of a bug in the venice provider in discord

https://discord.com/channels/1287729918100246654/1287729920319033345/1502127178651799767

Are you available to join the discussion and help look into it?
